### ISSUE TITLE
conditional isRequired propTypes

### ISSUE THREAD

**Issue by arizzitano** [edx/paragon#66]
We currently use `isRequiredIf` from [react-proptype-conditional-require](https://github.com/evcohen/react-proptype-conditional-require) for conditional propType requirement. It's useful for when we want to require specific props based on other props -- for example, `StatusAlert` only requiring an `onClose` callback function if it actually *has* a close button.

However, this function has a few drawbacks. As noted on #51, it [only works within the object scope where it's invoked](https://github.com/edx/paragon/pull/51/files#r146942783), meaning props nested within shapes or arrays don't have access to top-level props. For example, if a table is sortable, we might want to require specific props on its `headings` array, but can't do this because the `headings` array doesn't have access to the top-level `isSortable` prop.

To fix this, we could go one of three ways:
- Patch react-proptype-conditional-require upstream. The maintainer seems open to PRs. Once changes are merged, bump its version in Paragon.
- Use a different package, like [react-required-if](https://github.com/thejameskyle/react-required-if)
- Implement our own custom validator function as documented in [prop-types](https://github.com/facebook/prop-types)

**Comment by jaebradley** [comment]
### What happens to prop types checking inside a `shape`?

To understand what's happening with respect to a prop inside a `shape` that's referencing a prop outside that same `shape` (as mentioned in [this comment](https://github.com/edx/paragon/pull/51/files#r146942783)) we need to inspect [the `shape` prop type validator in the `prop-types` source code](https://github.com/facebook/prop-types/blob/01634d92e4e4cb5ffaf7de96ea8cb5da9b65e4f0/factoryWithTypeCheckers.js#L360-L380).

```javascript
function createShapeTypeChecker(shapeTypes) {
  function validate(props, propName, componentName, location, propFullName) {
    var propValue = props[propName];
    var propType = getPropType(propValue);
    if (propType !== 'object') {
      return new PropTypeError('Invalid ' + location + ' `' + propFullName + '` of type `' + propType + '` ' + ('supplied to `' + componentName + '`, expected `object`.'));
    }
    for (var key in shapeTypes) {
      var checker = shapeTypes[key];
      if (!checker) {
        continue;
      }
      var error = checker(propValue, key, componentName, location, propFullName + '.' + key, ReactPropTypesSecret);
      if (error) {
        return error;
      }
    }
    return null;
  }
  return createChainableTypeChecker(validate);
}
```

Ok, there's a lot of abstract-y things happening here. Basically, what I think is happening here is that when `PropTypes` tries to check a shape it

1. Gets the object for that `prop` (so in `Table` it would be the `column` prop)
2. Iterates through each property in the `shape` object
3. Calls the prop validator associated with the current `shape` property, passing it
    * the current `shape` object
    * the name of the prop, the name of the component
    * the "location" (in this case, `prop`)
    * a `string` "id" (in this case, it should be `column.columnSortable`)
    * and [`ReactPropTypesSecret`](https://github.com/facebook/prop-types/blob/01634d92e4e4cb5ffaf7de96ea8cb5da9b65e4f0/lib/ReactPropTypesSecret.js#L10) (which, I found out about today - kind've an Easter egg!)
4. And then does the standard `PropTypes` returning of an error if it exists

### Ok...maybe that's how it works - show me

We can see this in action with [an example I cooked up on `StackBlitz`](https://shape-prop-types.stackblitz.io)

```javascript
import React, { Component } from 'react';
import PropTypes from 'prop-types';

// Basic custom prop validator that will always return false / invalid
const customPropValidator = (props, propName, componentName, ...rest) => {
  console.log('Props passed to the custom email prop validator: ', props);
  console.log('Prop Name: ', propName);
  console.log('Component Name: ', componentName);
  console.log('Rest: ', rest);

  // always returns invalid
  return false;
};

class Test extends Component {
  constructor(props) {
    super(props);
  }

  render() {
    return (
      <div>
        Foobar
      </div>
    )
  }
}

Test.defaultProps = {
  details: {
    emailAddress: '',
    someOtherThing: 'some other thing',
  },
  showDetails: true,
}


Test.propTypes = {
  details: PropTypes.shape({
    emailAddress: customPropValidator,
    someOtherThing: PropTypes.string,
  }),
  showDetails: PropTypes.bool,
};

export default Test;
```

Note how I'm `console.log`ging in the custom prop validator.

So now, when we go to [the page](https://shape-prop-types.stackblitz.io/) and check the console...

![alt-text](https://i.imgur.com/dBCBvaP.png)

The `props` passed to the custom validator inside a `shape` are the `shape`'s keys, as the source code seems to indicate.

**TL;DR: The behavior we're seeing does not stem from `isRequiredIf` but rather how `shape` prop type validation has been implemented**

**Comment by joshuapinter** [comment]
Hey guys, I stumbled on this while looking for a solution to conditionally require a prop. I never liked the syntax of any of the helper packages available so the [company I work for][2] forked `prop-types` and made a `isRequiredIf` chained command that is built right into `prop-types`. Your issue goes a little beyond this but it might solve it as well.

So now you can do something like this: 

```js
MyComponent.propTypes = {
  firstProp: PropTypes.string.isRequiredIf( props => !props.otherProp ), // required only if `otherProp` not provided
  otherProp: PropTypes.string.isRequiredIf( props => !props.firstProp ), // required only if `firstProp` not provided
}
```

Super clean and minimal. 

Feel free to use [our fork][3] in your own project by updating your `package.json` with the following: 

```json
"prop-types": "github:cntral/prop-types#isRequiredIf"
```

_NOTE: It does not take a boolean param, only a function that is passed the props and needs to return a boolean._


  [1]: https://github.com/facebook/react/pull/5933
  [2]: https://www.cntr.al/
  [3]: https://github.com/cntral/prop-types/tree/isRequiredIf
